api/v1beta1/target.go

```diff
@@ -6,111 +6,252 @@ import (
+		}
+
+		// Try a pointer receiver
+		item, ok = itemsField.Index(i).Addr().Interface().(client.Object)
+		if ok {
+			items[i] = item
+
+			continue
+		}
+
+		return nil, &ReflectiveResourceListError{
+			typeName: value.Type().PkgPath() + "." + value.Type().Name(),
+			message: "an item in the underlying struct's 'Items' slice could not be " +
+				"type-asserted to a sigs.k8s.io/controller-runtime/pkg/client.Object",
+		}
+	}
+
+	return items, nil
+}
+
+func (l *ReflectiveResourceList) ObjectList() client.ObjectList {
+	return l.ClientList
+}
+
+//+kubebuilder:object:generate=false
+
+type ReflectiveResourceListError struct {
+	typeName string
+	message  string
+}
+
+func (e *ReflectiveResourceListError) Error() string {
+	return fmt.Sprintf("unable to use %v as a nucleus ResourceList: %v", e.typeName, e.message)
+}
 
-		listOpts.LabelSelector = labelSel
+// GetMatches returns a list of resources on the cluster, matched by the Target. The provided
+// ResourceList should be backed by a client.ObjectList type which must registered in the scheme of
+// the client.Reader. The items in the provided ResourceList after this method is called will not
+// necessarily equal the items matched by the Target.
+//
+// This method should be used preferentially to `GetMatchesDynamic` because it can leverage the
+// Reader's cache. NOTE: unlike the NamespaceSelector, an empty Target will match *all* resources on
+// the cluster.
+func (t Target) GetMatches(ctx context.Context, r client.Reader, list ResourceList) ([]client.Object, error) {
+	nonNilSel := t.LabelSelector
+	if nonNilSel == nil { // override it to be empty if it is nil
+		nonNilSel = &metav1.LabelSelector{}
 	}
 
-	namespaceList := &corev1.NamespaceList{}
-	if err := r.List(ctx, namespaceList, &listOpts); err != nil {
+	labelSel, err := metav1.LabelSelectorAsSelector(nonNilSel)
+	if err != nil {
 		return nil, err
 	}
 
-	namespaces := make([]string, len(namespaceList.Items))
-	for i, ns := range namespaceList.Items {
-		namespaces[i] = ns.GetName()
+	listOpts := client.ListOptions{
+		LabelSelector: labelSel,
+		Namespace:     t.Namespace,
+	}
+
+	if err := r.List(ctx, list.ObjectList(), &listOpts); err != nil {
+		return nil, err
 	}
 
-	return Target(sel).matches(namespaces)
+	items, err := list.Items()
+	if err != nil {
+		return nil, err
+	}
+
+	return t.matchesByName(items)
 }
 
-type Target struct {
-	*metav1.LabelSelector `json:",inline"`
+// GetMatchesDynamic returns a list of resources on the cluster, matched by the Target. The kind of
+// the resources is configured by the provided dynamic.ResourceInterface. If the Target specifies a
+// namespace, this method will limit the namespace of the provided Interface if possible. If the
+// provided Interface is already namespaced, the namespace of the Interface will be used (possibly
+// overriding the namespace specified in the Target).
+//
+// NOTE: unlike the NamespaceSelector, an empty Target will match *all* resources on the cluster.
```

Comment:
This is something I need to reconsider

Reply:
Careful documentation will be the best approach. The behavior of selectors is often confusing to me (and I've read through it on various occasions). Nil selectors might select nothing? But empty selectors select everything. It's like they're deliberately trying to confuse me. So we will treat nil as empty, and if a user wants to select nothing, they can `Exclude: "*"`.